solr/core/src/test/org/apache/solr/TestDistributedGrouping.java

```diff
@@ -425,6 +426,65 @@ public void test() throws Exception {
     
     //Debug
     simpleQuery("q", "*:*", "rows", 10, "fl", "id," + i1, "group", "true", "group.field", i1, "debug", "true");
+    doTestGroupSkipSecondStep();
+  }
+
+  /*
+    SOLR-11831, test skipping the second grouping step if the query only retrieves on document per group
+   */
+  private void doTestGroupSkipSecondStep() throws Exception {
+    ignoreException(GroupParams.GROUP_SKIP_DISTRIBUTED_SECOND); // don't print stack trace for exception raised by group.skip.second.step
+    // Ignore numFound if group.skip.second.step is enabled because the number of documents per group will not be computed (will default to 1)
+    handle.put("numFound", SKIP);
+    query("q", "{!func}id_i1", "rows", 3, "group.skip.second.step", true, "group.limit", 1,  "fl",  "id," + i1, "group", "true",
+        "group.field", i1);
+    query("q", "kings", "group.skip.second.step", true, "fl", "id," + i1, "group", "true", "group.field", i1);
+    query("q", "{!func}id_i1", "rows", 3, "group.skip.second.step", true,  "fl",  "id," + i1, "group", "true",
+        "group.field", i1);
+    query("q", "1234doesnotmatchanything1234", "group.skip.second.step", true, "fl", "id," + i1, "group", "true", "group.field", i1);
+
+    ignoreException("Illegal grouping specification");
+    // ngroups will return the corrent results, the problem is that numFound for each group might be wrong in case of multishard setting - but there is no way to
+    // enable/disable it.
+    //assertSimpleQueryThrows("q", "{!func}id_i1", "group.skip.second.step", true, "fl", "id," + i1, "group", "true", "group.field", i1, "group.ngroups", true);
```

Comment:
The commented upon and commented out test here jumped out. How about disallowing `group.ngroups=true` when `group.skip.second.step=true` is used? In the multi-shard case the numFound values for each group would be wrong, in the single-shard case the numFound values would be right but then in a single-sharded setup the usage of distributed grouping would be less likely (though not impossible) presumably? https://github.com/cpoerschke/lucene-solr/commit/f3d715c5b0ea708c15cace8f889e12e48292d79b has a potential code change (but no documentation edit yet).

Reply:
Merged, but I'm now thinking that if you are using only one shard per replica then `ngroups` will be right, would be nice to allow ngroups in that case, what do you think?